# **Data Preparation**

### **1. Install Necessary Libraries**

In [5]:
%pip install datasets huggingface_hub

Note: you may need to restart the kernel to use updated packages.


In [12]:
# Importing necessary libraries

import pandas as pd
from datasets import load_dataset

print("datasets library imported successfully")

# Importing train_test_split

from sklearn.model_selection import train_test_split

print("train_test_split imported successfully")

datasets library imported successfully
train_test_split imported successfully


### **2. Load Dataset from Hugging Face Repository**

In [2]:
# Loading dataset directly from Hugging Face

dataset = load_dataset(
    "gkrishna4346/predictive-maintenance-dataset"
)

# Display dataset structure

dataset

README.md:   0%|          | 0.00/834 [00:00<?, ?B/s]

C:\Users\HP\anaconda3\Install\Lib\site-packages\huggingface_hub\file_download.py:138: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\HP\.cache\huggingface\hub\datasets--gkrishna4346--predictive-maintenance-dataset. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


engine_data.csv:   0%|          | 0.00/1.30M [00:00<?, ?B/s]

Generating train split: 0 examples [00:00, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['Engine rpm', 'Lub oil pressure', 'Fuel pressure', 'Coolant pressure', 'lub oil temp', 'Coolant temp', 'Engine Condition'],
        num_rows: 19535
    })
})

### **3. Convert Dataset to Pandas DataFrame**

In [3]:
# Converting Hugging Face dataset into pandas DataFrame

df = dataset["train"].to_pandas()

# Display dataset dimensions

print("Dataset Shape:", df.shape)

# Display first 5 records

df.head()

Dataset Shape: (19535, 7)


,Engine rpm,Lub oil pressure,Fuel pressure,Coolant pressure,lub oil temp,Coolant temp,Engine Condition
0,700,2.493592,11.790927,3.178981,84.144163,81.632187,1
1,876,2.941606,16.193866,2.464504,77.640934,82.445724,0
2,520,2.961746,6.553147,1.064347,77.752266,79.645777,1
3,473,3.707835,19.510172,3.727455,74.129907,71.774629,1
4,619,5.672919,15.738871,2.052251,78.396989,87.000225,0


### **4. Data Cleaning and Verification**

#### **4.1. Check Data Types**

In [5]:
# Displaying column data types

df.dtypes

Engine rpm            int64
Lub oil pressure    float64
Fuel pressure       float64
Coolant pressure    float64
lub oil temp        float64
Coolant temp        float64
Engine Condition      int64
dtype: object

#### **4.2. Check Missing Values**

In [6]:
# Checking missing values

df.isnull().sum()

Engine rpm          0
Lub oil pressure    0
Fuel pressure       0
Coolant pressure    0
lub oil temp        0
Coolant temp        0
Engine Condition    0
dtype: int64

#### **4.3. Check Duplicate Records**

In [7]:
# Checking duplicate records

df.duplicated().sum()

0

**Observations:**

- The dataset contains 19,535 records and 7 columns.

- Data type verification confirms that all features are stored in appropriate numerical formats suitable for machine learning analysis.

- No missing values were identified across any feature.

- No duplicate records were detected in the dataset.

- All columns are relevant to the predictive maintenance problem statement and therefore no columns were removed.

- The dataset is clean, consistent, and ready for further preprocessing and model development.

### **5. Feature and Target Separation**

In [9]:
# Separating Features and Target Variable

X = df.drop("Engine Condition", axis=1)

y = df["Engine Condition"]

# Display dimensions

print("Features Shape:", X.shape)
print("Target Shape:", y.shape)

Features Shape: (19535, 6)
Target Shape: (19535,)


**Observations:**

- The target variable `Engine Condition` was separated from the feature set for supervised machine learning.

- The feature matrix contains 19,535 records and 6 independent variables.

- The target variable contains 19,535 corresponding labels representing engine condition classes.

- The dataset is now ready for train-test splitting and model preparation.

### **6. Train-Test Split**

In [13]:
# Splitting dataset into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

# Display split dimensions

print("X_train Shape:", X_train.shape)
print("X_test Shape :", X_test.shape)
print("y_train Shape:", y_train.shape)
print("y_test Shape :", y_test.shape)

X_train Shape: (15628, 6)
X_test Shape : (3907, 6)
y_train Shape: (15628,)
y_test Shape : (3907,)


### **7. Save Train and Test Datasets Locally**

In [14]:
# Combining features and target for training dataset

train_df = X_train.copy()
train_df["Engine Condition"] = y_train

# Combining features and target for testing dataset

test_df = X_test.copy()
test_df["Engine Condition"] = y_test

# Saving datasets locally

train_df.to_csv("../data/train_data.csv", index=False)
test_df.to_csv("../data/test_data.csv", index=False)

print("Training and testing datasets saved successfully")

Training and testing datasets saved successfully


In [15]:
# Verifying saved dataset dimensions

print("Train Dataset Shape:", train_df.shape)
print("Test Dataset Shape :", test_df.shape)

Train Dataset Shape: (15628, 7)
Test Dataset Shape : (3907, 7)


**Observations:**

- The cleaned dataset was successfully split into training and testing datasets using an 80:20 ratio.

- The feature variables and target variable were combined back together before saving.

- The training dataset contains 15,628 records, while the testing dataset contains 3,907 records.

- Both datasets were saved locally for subsequent modeling and evaluation activities.

### **8. Upload Train and Test Datasets to Hugging Face Data Space**

In [17]:
from huggingface_hub import list_repo_files

files = list_repo_files(
    repo_id="gkrishna4346/predictive-maintenance-dataset",
    repo_type="dataset"
)

print("Files available in Hugging Face Dataset Repository:\n")

for file in files:
    print(file)

Files available in Hugging Face Dataset Repository:

.gitattributes
README.md
engine_data.csv
test_data.csv
train_data.csv


### **9. Data Preparation Summary**

**Observations:**

- The dataset was successfully loaded from the Hugging Face data repository.

- Data cleaning and verification confirmed that the dataset contains no missing values, duplicate records, or unnecessary columns.

- The target variable was separated from the feature set to prepare the dataset for supervised machine learning.

- The cleaned dataset was split into training and testing datasets using an 80:20 ratio while preserving class distribution through stratified sampling.

- The resulting train and test datasets were saved locally and uploaded back to the Hugging Face repository for reproducibility and centralized access.

- The dataset is now fully prepared for feature scaling, model development, and performance evaluation.